crates/apollo-router-core/src/schema.rs

```diff
@@ -17,7 +17,10 @@ impl std::str::FromStr for Schema {
         let tree = parser.parse();
 
         if !tree.errors().is_empty() {
-            return Err(SchemaError::ParseErrors(tree.errors().to_vec()));
+            return Err(SchemaError::from_parse_errors(
+                s.to_string(),
+                tree.errors().to_vec(),
```

Comment:
Ahhh! It's cool that `errors()` finally returns a slice. Now you need to make `from_parse_errors()` accept a slice and do the `to_vec()` inside that function. This is because you want the vec to be a detail of implementation of your struct.

I think ideally `errors()` should return an `impl Iterator` (or IntoIterator, I always forget which one) and `from_parse_errors()` should also take that to make it really agnostic of the kind of collection that are behind the objects.

Reply:
Hmm, I tend to explicitly ask for owned versions in order to "not lie to the caller" about the possible allocations / ownership requirements.

I really dig the into_iterator < Item = Error > idea, let's see what it would look like! 🚀

Reply:
> Hmm, I tend to explicitly ask for owned versions in order to "not lie to the caller" about the possible allocations / ownership requirements.

I guess that makes sense :thinking: ..... though if you create a new object from a constructor and it takes references in arguments, you kinda expect there will be allocations XD

> I really dig the into_iterator < Item = Error > idea, let's see what it would look like! rocket

There is the example of FromIterator in std: https://doc.rust-lang.org/std/iter/trait.FromIterator.html

(This is the one you use when you do `.collect()`)

Reply:
ast.errors() should totally return an Iterator -- if y'all have the time to PR that in, I'd be very grateful 🙏

Reply:
Yay an opportunity for my first apollo-parser contribution :D